File: metadata/mill-metadata-autoconfigure/src/main/java/io/qpointz/mill/metadata/configuration/MetadataCoreConfiguration.java
package io.qpointz.mill.metadata.configuration;

import io.qpointz.mill.metadata.domain.*;
import io.qpointz.mill.metadata.domain.core.*;
import io.qpointz.mill.metadata.repository.FacetTypeRepository;
import io.qpointz.mill.metadata.repository.InMemoryFacetTypeRepository;
import io.qpointz.mill.metadata.service.*;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.util.Set;

@Slf4j
@Configuration
public class MetadataCoreConfiguration {

    @Bean
    @ConditionalOnMissingBean
    public FacetClassResolver facetClassResolver() {
        var resolver = new DefaultFacetClassResolver();
        resolver.register("structural", StructuralFacet.class);
        resolver.register("descriptive", DescriptiveFacet.class);
        resolver.register("relation", RelationFacet.class);
        resolver.register("concept", ConceptFacet.class);
        resolver.register("value-mapping", ValueMappingFacet.class);
        log.info("Registered platform facet class resolver with 5 mappings");
        return resolver;
    }

    @Bean
    @ConditionalOnMissingBean
    public FacetTypeRepository facetTypeRepository() {
        var repo = new InMemoryFacetTypeRepository();
        registerPlatformFacetTypes(repo);
        return repo;
    }

    @Bean
    @ConditionalOnMissingBean
    public FacetCatalog facetCatalog(FacetTypeRepository facetTypeRepository,
                                     @Autowired(required = false) FacetContentValidator contentValidator) {
        log.info("Creating DefaultFacetCatalog (content validator: {})",
                contentValidator != null ? "enabled" : "disabled");
        return new DefaultFacetCatalog(facetTypeRepository, contentValidator);
    }

    private void registerPlatformFacetTypes(FacetTypeRepository repo) {
        repo.save(FacetTypeDescriptor.builder()
                .typeKey("structural")
                .mandatory(true)
                .enabled(true)
                .displayName("Structural")
                .description("Physical schema binding")
                .applicableTo(Set.of(MetadataTargetType.TABLE, MetadataTargetType.ATTRIBUTE))
                .version("1.0")
                .build());

        repo.save(FacetTypeDescriptor.builder()
                .typeKey("descriptive")
                .mandatory(true)
                .enabled(true)
                .displayName("Descriptive")
                .description("Human-readable metadata")
                .applicableTo(Set.of(MetadataTargetType.SCHEMA, MetadataTargetType.TABLE, MetadataTargetType.ATTRIBUTE))
                .version("1.0")
                .build());

        repo.save(FacetTypeDescriptor.builder()
                .typeKey("relation")
                .mandatory(true)
                .enabled(true)
                .displayName("Relation")
                .description("Cross-entity relationships")
                .applicableTo(Set.of(MetadataTargetType.TABLE))
                .version("1.0")
                .build());

        repo.save(FacetTypeDescriptor.builder()
                .typeKey("concept")
                .mandatory(false)
                .enabled(true)
                .displayName("Concept")
                .description("Business concept definitions")
                .applicableTo(Set.of(MetadataTargetType.CONCEPT))
                .version("1.0")
                .build());

        repo.save(FacetTypeDescriptor.builder()
                .typeKey("value-mapping")
                .mandatory(false)
                .enabled(true)
                .displayName("Value Mapping")
                .description("Attribute value mappings for NL2SQL")
                .applicableTo(Set.of(MetadataTargetType.ATTRIBUTE))
                .version("1.0")
                .build());

        log.info("Registered 5 platform facet type descriptors (3 mandatory, 2 optional)");
    }
}

File: metadata/mill-metadata-autoconfigure/src/main/kotlin/io/qpointz/mill/metadata/configuration/MetadataCoreConfiguration.kt
package io.qpointz.mill.metadata.configuration

import io.qpointz.mill.metadata.repository.FacetTypeDefinitionRepository
import io.qpointz.mill.metadata.repository.FacetTypeRepository
import io.qpointz.mill.metadata.repository.InMemoryFacetTypeDefinitionRepository
import io.qpointz.mill.metadata.repository.InMemoryFacetTypeRepository
import io.qpointz.mill.metadata.service.DefaultFacetCatalog
import io.qpointz.mill.metadata.service.FacetCatalog
import org.slf4j.LoggerFactory
import org.springframework.boot.ApplicationRunner
import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

/**
 * Core metadata beans: facet catalog and registry guards.
 *
 * **Seeding:** platform scopes, facet types, and entities are applied **only** through
 * `mill.metadata.seed.resources` ([MetadataSeedStartup]); there are no Flyway data inserts or
 * autoconfigure `ApplicationRunner` seeders for facet definitions.
 */
@Configuration
@EnableConfigurationProperties(MetadataProperties::class)
open class MetadataCoreConfiguration {

    /**
     * In-memory facet type definitions when JPA metadata storage is not active.
     *
     * <p>Skipped when {@code mill.metadata.repository.type=jpa} (JPA adapters) or {@code file}
     * ([MetadataFileRepositoryAutoConfiguration] supplies in-memory file-backed repos).
     */
    @Bean
    @ConditionalOnMissingBean(FacetTypeDefinitionRepository::class)
    @ConditionalOnExpression(
        "'\${mill.metadata.repository.type:file}' != 'jpa' && '\${mill.metadata.repository.type:file}' != 'file' " +
            "&& '\${mill.metadata.facet-type-registry.type:inMemory}' == 'inMemory'"
    )
    open fun inMemoryFacetTypeDefinitionRepository(): FacetTypeDefinitionRepository =
        InMemoryFacetTypeDefinitionRepository()

    /**
     * In-memory runtime facet type rows paired with [inMemoryFacetTypeDefinitionRepository].
     */
    @Bean
    @ConditionalOnMissingBean(FacetTypeRepository::class)
    @ConditionalOnExpression(
        "'\${mill.metadata.repository.type:file}' != 'jpa' && '\${mill.metadata.repository.type:file}' != 'file' " +
            "&& '\${mill.metadata.facet-type-registry.type:inMemory}' == 'inMemory'"
    )
    open fun inMemoryFacetTypeRepository(): FacetTypeRepository = InMemoryFacetTypeRepository()

    /**
     * Creates the default [FacetCatalog] backed by definition and runtime type repositories.
     *
     * @param definitionRepository persisted or in-memory facet type definitions
     * @param facetTypeRepository runtime facet type rows
     * @return configured catalog instance
     */
    @Bean
    @ConditionalOnMissingBean(FacetCatalog::class)
    open fun facetCatalog(
        definitionRepository: FacetTypeDefinitionRepository,
        facetTypeRepository: FacetTypeRepository
    ): FacetCatalog {
        log.info("Creating DefaultFacetCatalog")
        return DefaultFacetCatalog(definitionRepository, facetTypeRepository)
    }

    /**
     * Validates facet type registry strategy selection at startup.
     *
     * `portal` is declared as a contract for future work and intentionally fails fast.
     * `local` expects local persistence-backed descriptor sourcing (typically JPA).
     */
    @Bean
    open fun facetTypeRegistryStrategyGuard(
        properties: MetadataProperties
    ): ApplicationRunner = ApplicationRunner {
        val strategy = properties.facetTypeRegistry.type
        when (strategy) {
            "portal" -> throw IllegalStateException(
                "mill.metadata.facet-type-registry.type=portal is not implemented yet"
            )
            "local" -> {
                if (properties.repository.type != "jpa") {
                    throw IllegalStateException(
                        "mill.metadata.facet-type-registry.type=local requires mill.metadata.repository.type=jpa"
                    )
                }
            }
            "inMemory" -> {
                // default fallback strategy
            }
            else -> throw IllegalStateException(
                "Unsupported mill.metadata.facet-type-registry.type='$strategy'. Supported: inMemory, local, portal"
            )
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(MetadataCoreConfiguration::class.java)
    }
}
